security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -0,0 +1,195 @@
+	closing        chan bool
+
+	// Cached secret will be removed from cache if (time.now - secretItem.lastUsedTime >= evictionDuration), this prevents cache growing indefinitely.
+	evictionDuration time.Duration
+
+	// Key rotation job running interval.
+	rotationInterval time.Duration
+
+	// secret TTL.
+	secretTTL time.Duration
+
+	// How may times that key rotation job has detected secret change happened, used in unit test.
+	secretChangedCount uint64
+}
+
+// NewSecretCache creates a new secret cache.
+func NewSecretCache(cl CAClient, secretTTL, rotationInterval, evictionDuration time.Duration) *SecretCache {
+	ret := &SecretCache{
+		caClient:         cl,
+		rotationInterval: rotationInterval,
+		evictionDuration: evictionDuration,
+		secretTTL:        secretTTL,
+		closing:          make(chan bool, 1),
+	}
+
+	atomic.StoreUint64(&ret.secretChangedCount, 0)
+	go ret.keyRotationJob()
+	return ret
+}
+
+// GetSecret gets secret from cache.
+func (sc *SecretCache) GetSecret(proxyID, jwtToken string) (*SecretItem, error) {
+	now := time.Now()
+
+	// Return directly if secret could be found from cache and not expired.
+	val, found := sc.secrets.Load(proxyID)
+	if found {
+		s := val.(SecretItem)
+		if s.jwtToken == jwtToken && !sc.checkExpired(&s) {
+			// Update cached secret's lastUsedTime.
+			s.lastUsedTime = now
+			return &s, nil
+		}
+	}
+
+	ns, err := sc.generateSecret(jwtToken, now)
+	if err != nil {
+		log.Errorf("Failed to generate secret for proxy %q: %v", proxyID, err)
+		return nil, err
+	}
+
+	sc.secrets.Store(proxyID, *ns)
+	return ns, nil
+}
+
+// Close shuts down the secret cache.
+func (sc *SecretCache) Close() {
+	sc.closing <- true
+}
+
+func (sc *SecretCache) keyRotationJob() {
+	// Wake up once in a while and refresh stale items.
+	sc.rotationTicker = time.NewTicker(sc.rotationInterval)
+	for {
+		select {
+		case now := <-sc.rotationTicker.C:
+			sc.rotate(now)
+		case <-sc.closing:
+			sc.rotationTicker.Stop()
+			return
+		}
+	}
+}
+
+func (sc *SecretCache) rotate(t time.Time) {
+	sc.secrets.Range(func(key interface{}, value interface{}) bool {
+		proxyID := key.(string)
+		now := time.Now()
+
+		e := value.(SecretItem)
+
+		// Remove from cache if the secret hasn't been used for a while, this prevent the cache growing indefinitely.
+		if now.After(e.lastUsedTime.Add(sc.evictionDuration)) {
+			sc.secrets.Delete(proxyID)
+			return true
+		}
+
+		// Re-generate secret if it's expired.
+		if sc.checkExpired(&e) {
+			go func() {
+				ns, err := sc.generateSecret(e.jwtToken, now)
+				if err != nil {
+					log.Errorf("Failed to generate secret for proxy %q: %v", proxyID, err)
+					return
+				}
+
+				sc.secrets.Store(proxyID, *ns)
+
+				//TODO(quanlin): push new secret to proxy.
+
+				atomic.AddUint64(&sc.secretChangedCount, 1)
+			}()
+		}
+
+		return true
+	})
+}
+
+func (sc *SecretCache) generateSecret(jwtToken string, t time.Time) (*SecretItem, error) {
+	options := util.CertOptions{
+		Host:       jwtToken,
```

Comment:
jwtToken is a credential should not be the value of `Host`.

Since it's just a prototype, it's fine to leave it empty & todo.

Reply:
I remember Citadel uses service account as certoption.host, not sure if it applies here until integration story is clear, my understanding is as long as it's a unique value should be fine ? anyway, remove jwtToken and add TODO.

Reply:
Citadel is using SPIFFE id for the host, which includes the k8s service account.
https://github.com/istio/istio/blob/33f6c791bf44f4814ed34815b2ebe61020fb574d/security/pkg/pki/ca/controller/secret.go#L297

We need to make sure the `CertOptions.Host` and x509.SAN field represents the actual identity. In Google world, that corresponds to a GCP service account. Token is just to get CSR approved.

Reply:
I see, thanks !